apps/eth_blockchain/lib/eth_blockchain/transaction.ex

```diff
@@ -0,0 +1,261 @@
+         {:ok, nonce} <- get_transaction_count({from_address}, adapter, pid) do
+      gas_limit = Application.get_env(:eth_blockchain, :default_contract_transaction_gas_limit)
+
+      %__MODULE__{
+        gas_limit: gas_limit,
+        gas_price: gas_price,
+        nonce: int_from_hex(nonce),
+        to: from_hex(contract_address),
+        data: encoded_abi_data
+      }
+      |> sign_and_hash(from_address)
+      |> send_raw(adapter, pid)
+    else
+      error -> error
+    end
+  end
+
+  defp sign_and_hash(%__MODULE__{} = transaction_data, from_address) do
+    case sign_transaction(transaction_data, from_address) do
+      {:ok, signed_trx} ->
+        hashed =
+          signed_trx
+          |> serialize()
+          |> ExRLP.encode()
+          |> to_hex()
+
+        {:ok, hashed}
+
+      error ->
+        error
+    end
+  end
+
+  defp send_raw({:ok, transaction_data}, adapter, pid) do
+    Adapter.call(adapter, {:send_raw, transaction_data}, pid)
+  end
+
+  defp send_raw(error, _adapter, _pid), do: error
+
+  defp get_transaction_count({address}, adapter, pid) do
+    Adapter.call(adapter, {:get_transaction_count, address}, pid)
+  end
+
+  defp sign_transaction(transaction, wallet_address) do
+    chain_id = Application.get_env(:eth_blockchain, :chain_id)
+
+    result =
+      transaction
+      |> transaction_hash(chain_id)
+      |> Signature.sign_transaction_hash(wallet_address, chain_id)
+
+    case result do
+      {:ok, {v, r, s}} -> {:ok, %{transaction | v: v, r: r, s: s}}
+      error -> error
+    end
+  end
+
+  @doc """
+  Serialize, encode and returns a hash of a given transaction
+  """
+  @spec transaction_hash(__MODULE__.t(), integer()) :: Keccak.keccak_hash()
+  def transaction_hash(trx, chain_id) do
+    trx
+    |> serialize(false)
+    |> Kernel.++([encode_unsigned(chain_id), <<>>, <<>>])
+    |> ExRLP.encode()
+    |> Keccak.kec()
+  end
+
+  @doc """
+  Encodes a transaction such that it can be RLP-encoded.
+  """
+  @spec serialize(__MODULE__.t(), bool()) :: ExRLP.t()
+  def serialize(trx, include_vrs \\ true) do
+    base = [
+      trx.nonce |> encode_unsigned(),
+      trx.gas_price |> encode_unsigned(),
+      trx.gas_limit |> encode_unsigned(),
+      trx.to,
+      trx.value |> encode_unsigned(),
+      if(trx.to == <<>>, do: trx.init, else: trx.data)
+    ]
+
+    if include_vrs do
+      base ++
+        [
+          trx.v |> encode_unsigned(),
+          trx.r |> encode_unsigned(),
+          trx.s |> encode_unsigned()
+        ]
+    else
+      base
+    end
+  end
+
+  @doc """
+  Decodes a transaction that was previously encoded using `Transaction.serialize/1`.
+  """
+  @spec deserialize(ExRLP.t()) :: __MODULE__.t()
+  def deserialize(rlp) do
+    [
+      nonce,
+      gas_price,
+      gas_limit,
+      to,
+      value,
+      init_or_data,
+      v,
+      r,
+      s
+    ] = rlp
+
+    {init, data} = if to == <<>>, do: {init_or_data, <<>>}, else: {<<>>, init_or_data}
```

Comment:
🚨👮  **THIS IS `if` POLICE** 👮‍♀ 🚔  
YOU MUST SHOW YOUR ERLANG LICENSE

----------

Please use case 😢 

```
{init, data} =
  case to do
    <<>> -> {init_or_data, <<>>}
    _    -> {<<>>, init_or_data}
  end
```